automation_tools/satellite6/capsule.py

```diff
@@ -0,0 +1,398 @@
+
+    """
+    def __init__(self, user=None, password=None, key_filename=None):
+        self.user = user
+        self.password = password
+        self.key_filename = key_filename
+
+
+class HostConfig(Credentials):
+    """Stores a host's hostname and credentials information.
+
+    Usage::
+
+        cred1 = HostConfig('host1.example.com', 'admin', 'password')
+        cred2 = HostConfig(
+            'host2.example.com', key_filename='/path/to/ssh.key')
+
+    """
+    def __init__(self, hostname=None, port=22, *args, **kwargs):
+        super(HostConfig, self).__init__(*args, **kwargs)
+        self.hostname = hostname
+        self.port = port
+
+    @property
+    def host_string(self):
+        """Return a host_string in the format expected by Fabric"""
+        return '{}@{}:{}'.format(self.user, self.hostname, self.port)
+
+
+class Config(object):
+    """Configuration information provide easy access to configuration and some
+    helper methods to identify is some configuration is present or not.
+
+    """
+    def __init__(self, path):
+        self.path = path
+        self.organization_label = None
+        self.environment = None
+        self.content_view = None
+        self.activation_key = None
+        self.admin_user = None
+        self.admin_password = None
+        self.defaults = None
+        self.server = None
+        self.capsules = []
+        self._key_filenames = set()
+
+        self._parse()
+
+    def _parse(self):
+        """Parse the configuration and store the contents"""
+        with open(self.path) as handler:
+            data = json.load(handler)
+        self.organization_label = data.get('organization-label')
+        self.environment = data.get('environment')
+        self.content_view = data.get('content-view')
+        self.activation_key = data.get('activation-key')
+        self.admin_user = data.get('admin-user')
+        self.admin_password = data.get('admin-password')
+        defaults = data.get('defaults')
+        if defaults is not None and isinstance(defaults, dict):
+            key_filename = defaults.get('key-filename')
+            self._key_filenames.add(key_filename)
+            self.defaults = Credentials(
+                user=defaults.get('user'),
+                password=defaults.get('password'),
+                key_filename=key_filename,
+            )
+        server = data.get('server')
+        if server is not None and isinstance(server, dict):
+            key_filename = server.get('key-filename')
+            self._key_filenames.add(key_filename)
+            self.server = HostConfig(
+                hostname=server.get('hostname'),
+                user=server.get('user', self.defaults.user),
+                password=server.get('password', self.defaults.password),
+                key_filename=key_filename,
+            )
+        capsules = data.get('capsules')
+        if capsules is not None and isinstance(capsules, list):
```

Comment:
Hmm. I suspect there's a better way to handle this. But your explanation is plausible and I don't have a better solution in hand, so OK.